### DISCUSSION TITLE
How to implement a controller for a widget with a form

### DISCUSSION THREAD

**Discussion by ovekaaven** [OrchardCMS/OrchardCore#15102] [Q&A]
It would be nice to have documentation, or an example, for how to do the following:

Part 1 (easy enough): a widget that contains a form written in Razor. (Orchard Core's own Forms module is too inflexible.)

Part 2 (the part I can't figure out): a controller implemented in C#, that would check the submitted form data and if there's a mistake, can send the user back to the original page with the form contents still intact, so the user can correct the mistake and try again. (Not that difficult if there's no need to remember the form contents, but there is, and I have no idea how to get the form data back to the page.)

Our website has been OC-based for close to a year and I still don't know how to do it. I once tried to look at the source code of the Forms module to see how it does it, but it was a total mess and I couldn't follow it.

**Comment by MikeAlhayek** [comment]
@ovekaaven 
> Orchard Core's own Forms module is too inflexible.

Maybe list what you would like to accomplish here so we know the limitation.

> Part 2 (the part I can't figure out)
Are you using Workflow to handle the submitting logic? If you, I recently implemented a workflow task that does what you seems to be after `Http Redirect To Form Location Task`. So when the validation failed, you can just redirect back to the form.

**Reply by ovekaaven** [reply]
You're asking why I want to write the form in Razor? There are very few styling and layout options in Forms. It's difficult to make a tabular, multicolumn layout with just Forms. In any case, using Razor is supposed to be a supported use case.

No, I'm not using workflows, and so far have had no interest in that kind of massive overhead for tasks that can be done easier and more efficiently in C#. We do pay for the resources our service uses.

**Reply by MikeAlhayek** [reply]
@ovekaaven 

Don't forget that OC is written on the top of ASP.Net Core framework. So you are always welcome to add a razor view that posts to your own action.  So you can still do a traditional MVC code without using related to OC. You can also use shape and display drivers instead of using a single view model.

**Reply by ovekaaven** [reply]
Yes, that's pretty much what I've been doing for all my forms (Orchard Dojo has several examples of MVC code, e.g. [Controllers/YourFirstOrchardCoreController.cs](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Controllers/YourFirstOrchardCoreController.cs)). That's what I meant when, in my original post, I said I wanted a controller implemented in C#. Now could you help me with what I'm asking for?

**Comment by ns8482e** [comment]
Creating controller in orchard module is not different than creating controller in asp.net without orchard.

- create a controller for Get Action (Index) and return view for your model that renders html form
- create an action for POST request  and perform ModalState validation for your model and return same form view with model in case of model posted is invalid.
- If ModelState is valid, you can create/update content within your controller's post request by injecting and using IContentManager 

Check sample code in Demo module 

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Demo/Controllers/ContentController.cs

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Demo/Controllers/TodoController.cs

**Reply by ovekaaven** [reply]
That's pretty much what I've been doing so far, but only works if the form and the page is one and the same. So, again: if the form is part of a widget that's on some ordinary Orchard page (not itself a Razor page I've created, just an Orchard page with a flowpart with widgets, only one of which I want to define with a Razor view), and form validation fails (ModelState is invalid), how do you redisplay that original page without losing the form data (or model error message), so the user can correct their mistake. As a documentation feature request, I wanted an example of doing that be provided for clueless users like me, who wants to do this but can't quite figure out what kind of arcane plumbing needs to be done in order to get the form data unharmed back through Orchard's mysterious shape machinery and into the widget's Razor view.

**Reply by MikeAlhayek** [reply]
I don't think I understand your setup. 

So you have a content type with a flow part. On the flow part, you want to create a custom widget that renders a form?

**Reply by ovekaaven** [reply]
Yes, for the sake of discussion this could be the content type called "Page" from the standard recipes (although an alternative might be to put my custom widget into the zones of Design > Widgets in the admin menu). Making this widget is part 1, which, as stated, I know how to do. The problematic part is part 2, which is writing the controller for it, with a way to allow the whole page containing the widget to be redisplayed without loss of form data.

**Reply by MikeAlhayek** [reply]
@ovekaaven you don't want to mix your C# for with the content item edit form. 

If you want to create a form outside of workflow handling. You should create a form purely using C# and not relay on OC to handle anything for you. 

1. create a CustomFormPart with CustomFormPartDisplayDriver via C#. in your driver you can initialize a view that would render the form tag, your form input and the submit button. In the view you can add a hidden field that would give you the location of the current page. here is an example https://github.com/OrchardCMS/OrchardCore/blob/d783019615bf873cf2b5cab99b370ac934439d14/src/OrchardCore.Modules/OrchardCore.Forms/Views/Form.Wrapper.cshtml#L25
2. Register the part using services.AddContentPart<CustomFormPart>().UseDisplayDriver<CustomFormPartDisplayDriver>();
3. use a migration to create "CustomFormWidget" type with the Widget Stereotype and attach CustomFormPart part to the type using a migration.
4. using the UI, attach the `CustomFormWidget` using zones not as content part.
5. Now you should be able to submit the form. if the model state is invalid, then you return the user back to the location found in the hidden form value.

**Reply by ovekaaven** [reply]
> you don't want to mix your C# for with the content item edit form.

Nobody said I wanted that.

>     5. Now you should be able to submit the form. if the model state is invalid, then you return the user back to the location found in the hidden form value.

The devil's in the details, I suppose. Now, the problem I haven't been able to figure out for months, is how to do the "return the user back to the location" part, *without losing the form data*.

**Reply by rikbosch** [reply]
@ovekaaven if the widget can be rendered on any page you might want to post the form to an api method using javascript, and handle the validation errors in javascript as well or use something like htmx to render partial html.

If the user should be redirected after a successful form post, this should then also be handled clientside.

**Comment by ns8482e** [comment]
Form in razor page doesn't directly tied to your POST controller action. It's magic of MVC model binding.

You can create widget form using OC forms module and have the data posted to your controller POST action. As long as you define your model on controller action that MVC model binding understand, you will see your model properties get populated. You can add custom validation on your model.
